### PR TITLE
Feature/graceful shutdown

### PR DESCRIPTION
**Pull request by alpaka-alpachino**

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -30,6 +35,7 @@ func main() {
 	}
 
 	if err = consulConfig.ServiceRegister(consulClient, agentConfig); err != nil {
+		log.Println("there")
```
Comment (crnopster) [inline cmd/main.go]: Do we need this?

File: cmd/shutdown.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+	select {
+	case <-ctx.Done():
+		log.Println(ctx.Err())
+	case <-time.After(shutdownTimeout):
+		log.Println("Overslept")
+		cancel()
+	default:
+		cancel()
+	}
```
Comment (crnopster) [inline cmd/shutdown.go]: Can you explain logic of this?

Comment (alpaka-alpachino) [reply cmd/shutdown.go]: Think, I should use that for safety canceling when context expires. Case with Done() is selected when the context passed as an argument (in our occasion - "cancel") notifies completion, I also use use After() to insure we cancel if it takes too much time. But now I see, I don't need default.

File: cmd/main.go (hunk 1 above)
Comment (alpaka-alpachino) [reply cmd/main.go]: Well, not realy, thanks for your notice)

File: cmd/main.go [hunk 3]
```diff
@@ -21,19 +26,13 @@ func main() {
...
 	consulConfig := viper.NewConsulConfig()
-	agentConfig := consulConfig.AgentConfig()
 	consulClient, err := consulConfig.NewClient()
 
 	if err != nil {
 		log.Println(err)
 		return
 	}
 
-	if err = consulConfig.ServiceRegister(consulClient, agentConfig); err != nil {
```
Comment (dmitriyminer) [inline cmd/main.go]: Please pay attention to removed lines

Comment (alpaka-alpachino) [reply cmd/main.go]: Oh, yeah,thank you)

**Review by crnopster — DISMISSED** [review]

File: cmd/shutdown.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	timeout := make(chan bool)
+
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
+	sig := <-sigs
+	log.Println("Recieved", sig, "signal")
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	shutdownTime := time.Now().Add(shutdownTimeout)
```
Comment (dmitriyminer) [inline cmd/shutdown.go]: You can use `context.WithTimeout` instead of time variable

File: go.mod [hunk 5]
```diff
@@ -5,5 +5,6 @@ go 1.13
 require (
 	github.com/hashicorp/consul/api v1.4.0
 	github.com/lib/pq v1.3.0
+	github.com/segmentio/kafka-go v0.3.5
```
Comment (dmitriyminer) [inline go.mod]: I can't find usage of `github.com/segmentio/kafka-go`
Is it used?

File: cmd/shutdown_test.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+	"testing"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/connect"
+)
+
+func Test_gracefulShutdown_Successful(t *testing.T) {
+	signal := make(chan bool)
+
+	client, err := api.NewClient(api.DefaultConfig())
```
Comment (dmitriyminer) [inline cmd/shutdown_test.go]: It would be great to use io.Closer mock instead of consul client or db connection

File: cmd/shutdown.go [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+	"syscall"
+	"time"
+)
+
+const (
+	shutdownTimeout = 10 * time.Second
+)
+
+//GracefulShutdown implements releasing all resouces it got from system, finish all request handling and return responses when service stopping.
+func gracefulShutdown(done chan<- bool) error {
```
Comment (dmitriyminer) [inline cmd/shutdown.go]: Do you really need `done` channel if you close it inside function and waiting outside?

File: cmd/shutdown.go (hunk 4 above)
Comment (alpaka-alpachino) [reply cmd/shutdown.go]: fixed

File: cmd/shutdown.go (hunk 7 above)
Comment (alpaka-alpachino) [reply cmd/shutdown.go]: Chan used for blocking main until graceful shutdown will close all connections

File: go.mod (hunk 5 above)
Comment (alpaka-alpachino) [reply go.mod]: fixed

File: cmd/shutdown.go (hunk 7 above)
Comment (dmitriyminer) [reply cmd/shutdown.go]: So you can call `gracefulShutdown()` and wait

File: cmd/shutdown.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+	sigs := make(chan os.Signal)
+
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
```
Comment (dmitriyminer) [inline cmd/shutdown.go]: Signal and notify should be outside of this function.
Shutdown process starts after os.Signal syscall

File: cmd/shutdown.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+		// Please, replace srv with your server.
+		server := &http.Server{
+			Addr: ":8080",
+		}
+
+		for _, component := range components {
+			component.Close()
+		}
+
+		if err := server.Shutdown(ctx); err != nil {
```
Comment (dmitriyminer) [inline cmd/shutdown.go]: What about shutdown server first?

File: cmd/shutdown.go [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+			log.Fatalf("Server shutdown failed:%+v", err)
```
Comment (dmitriyminer) [inline cmd/shutdown.go]: Do you really need to exit with `log.Fatalf`?

File: cmd/shutdown.go [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+	"syscall"
+	"time"
+)
+
+const (
+	shutdownTimeout = 10 * time.Second
+)
+
+//GracefulShutdown implements releasing all resouces it got from system, finish all request handling and return responses when service stopping.
+func gracefulShutdown(done chan<- bool) error {
```
Comment (dmitriyminer) [inline cmd/shutdown.go]: Function returns error by signature but it doesn't return error anywhere

File: cmd/main.go [hunk 12]
```diff
@@ -44,10 +49,13 @@ func main() {
...
-	defer db.Close()
+	components = append(components,
+		//Put connection variables here
+		db)
 
-	for {
-		log.Println(" [INFO] App is running.")
-		time.Sleep(5 * time.Second)
-	}
+	gracefulShutdown(done)
```
Comment (dmitriyminer) [inline cmd/main.go]: Unhandled error

Comment (alpaka-alpachino) [reply cmd/main.go]: fixed

File: cmd/shutdown.go (hunk 11 above)
Comment (alpaka-alpachino) [reply cmd/shutdown.go]: fixed

File: cmd/shutdown.go (hunk 10 above)
Comment (alpaka-alpachino) [reply cmd/shutdown.go]: fixed

File: cmd/shutdown.go (hunk 9 above)
Comment (alpaka-alpachino) [reply cmd/shutdown.go]: nice idea))))))))))
fixed

File: cmd/shutdown.go (hunk 8 above)
Comment (alpaka-alpachino) [reply cmd/shutdown.go]: fixed

File: cmd/shutdown.go (hunk 7 above)
Comment (alpaka-alpachino) [reply cmd/shutdown.go]: yep

File: cmd/shutdown_test.go (hunk 6 above)
Comment (alpaka-alpachino) [reply cmd/shutdown_test.go]: fixed

**Review by crnopster — DISMISSED** [review]

File: cmd/shutdown.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+)
+
+//GracefulShutdown implements releasing all resouces it got from system, finish all request handling and return responses when service stopping.
+func gracefulShutdown(done chan<- bool, components []io.Closer) error {
+	ctx, cancel := context.WithTimeout(context.Background(), shutdownTimeout)
+
+	var errTimeout error
+
+	go func() {
+		for {
```
Comment (dmitriyminer) [inline cmd/shutdown.go]: I think you have high cpu usage in the goroutine, please check

File: cmd/shutdown.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+			errTimeout = ctx.Err()
```
Comment (dmitriyminer) [inline cmd/shutdown.go]: You can use ctx.Done() instead of the `for` loop. Two problems: race on `errTimeout` and high cpu usage

File: go.mod [hunk 15]
```diff
@@ -3,6 +3,7 @@ module github.com/lvl484/positioning-filter
 go 1.13
 
 require (
+	github.com/hashicorp/consul v1.7.2
```
Comment (dmitriyminer) [inline go.mod]: Unused library

Comment (alpaka-alpachino) [reply go.mod]: fixed

File: cmd/shutdown.go (hunk 14 above)
Comment (alpaka-alpachino) [reply cmd/shutdown.go]: fixed

**Review by crnopster — DISMISSED** [review]

File: cmd/shutdown_test.go [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+		expected: expected,
+	}
+}
+
+func (cm CloserMock) Close() error {
+	return cm.expected()
+}
+
+func TestGracefulShutdownSuccess(t *testing.T) {
+	timeout := time.Second * 10
```
Comment (i3odja) [inline cmd/shutdown_test.go]: You can use your constant shutdownTimeout there. It is the same value. It is unnecessary create new variable

Comment (alpaka-alpachino) [reply cmd/shutdown_test.go]: fixed

**Review by crnopster — DISMISSED** [review]

File: cmd/shutdown.go [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+import (
+	"context"
+	"io"
+	"time"
+)
+
+//GracefulShutdown implements releasing all resouces it got from system, finish all request handling and return responses when service stopping.
+func gracefulShutdown(timeout time.Duration, components []io.Closer) error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+
```
Comment (dmitriyminer) [inline cmd/shutdown.go:14]: defer cancel()

File: cmd/shutdown.go [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+	for _, component := range components {
+		select {
+		case <-ctx.Done():
+			err := ctx.Err()
+			cancel()
+			return err
```
Comment (dmitriyminer) [inline cmd/shutdown.go]: return ctx.Err()

Comment (alpaka-alpachino) [reply cmd/shutdown.go]: oh, okay, I wrote so because of linter, but I think your varient better too

File: cmd/shutdown.go [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+		default:
+			component.Close()
+			if err := component.Close(); err != nil {
+				return err
```
Comment (dmitriyminer) [inline cmd/shutdown.go]: log.Printf close failed and error, do no return on error, you should try to close all closers

**Review by dmitriyminer — DISMISSED** [review]

**Review by crnopster — DISMISSED** [review]

**Review by VasylKosiv — DISMISSED** [review]

File: cmd/shutdown.go [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	for _, component := range components {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			component.Close()
+			if err := component.Close(); err != nil {
```
Comment (dmitriyminer) [inline cmd/shutdown.go:20]: Close call two times

File: cmd/shutdown.go [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+				log.Printf("Cant close, err: %v", err)
+			}
+		}
+	}
+
+	cancel()
```
Comment (dmitriyminer) [inline cmd/shutdown.go]: Already cancelled with defer

File: cmd/shutdown.go (hunk 20 above)
Comment (alpaka-alpachino) [reply cmd/shutdown.go:20]: fixed

File: cmd/shutdown.go (hunk 21 above)
Comment (alpaka-alpachino) [reply cmd/shutdown.go]: fixed

**Review by dmitriyminer — APPROVED** [review]

**Review by i3odja — APPROVED** [review]

**Review by maksymbilokur — APPROVED** [review]

**Pull request merged**